### PR TITLE
Notification

### LINKED ISSUE #10: Add component : Notifications
As an MVP component, I suggest to provide some basic notifications:

* Duplicate the style of the Sidebar component
  * And use the same animation delays
  * On large screens, add borders to the borders that are not on the viewport edge
  * max-width: 500px
* Have a delay prop
  * 4000 ms by default
  * Can be set to 0 for no delay
* Have a position prop: top, top-right, right, etc.
* Have a direction prop: top, right, bottom, left (direction of the animation)
* Have an opened prop
* No Header/Footer handling but a simple title prop
* The content is handled by the children
* Two possible CTA props
  * No CTA: Display a cancel button
  * 1 CTA: Display the cancel button and the CTA
  * 2 CTA: Display the 2 CTA buttons
  * Set opened to false when clicking on a CTA
* No cross to close the notification

### PR DESCRIPTION
**Pull request by bleucitron**
Added `Notification` and `Notification.Group` components with basic tests.

Props are a bit different from what we discussed.
```
Notification.propTypes = {
  id: PropTypes.number.isRequired, // required to be able to use properly the NotificationCenter
  children: PropTypes.node.isRequired,
  onTimeoutClose: PropTypes.func, // passed by the NotificationCenter to unmount the Notification
  context: PropTypes.oneOf(contexts),
  position: PropTypes.oneOf([
    'top-left',
    'top-right',
    'bottom-right',
    'bottom-left',
    'top',
    'bottom'
  ]),
  direction: PropTypes.oneOf(['horizontal', 'vertical']),
  callToActions: PropTypes.array, // [{text, action}]
  pending: PropTypes.bool, // if set, the Notification will stay
  timespan: PropTypes.number, // better name than delay in my opinion
  className: PropTypes.string
};
```

I chose to let the developer define when a `Notification` should stay pending rather than make it depend on the number of `callToActions`.

You need to set up a `NotificationCenter` container (see `FakeNotificationCenter` in the stories) to control the unmounting of `Notifications`.
A `Notification` should always be used within a `Notification.Group`, but can be used alone if necessary.

Fixes #10.

### CODE REVIEW

**Comment by GuillaumeAmat** [comment]
I did not see the code for now but here are some glitches ^^:

peek 14-12-2017 20-28
peek 14-12-2017 20-29
peek 14-12-2017 20-30

File: src/index.js
Comment (GuillaumeAmat) [inline src/index.js]: Why did you remove that export?

File: stories/Modal.jsx
Comment (GuillaumeAmat) [inline stories/Modal.jsx]: Adding a 30px padding is better IMO.

File: src/components/Modal/Modal.jsx
Comment (GuillaumeAmat) [inline src/components/Modal/Modal.jsx:123]: The header element should not be displayed if there is no header prop.

File: src/components/Toolbar/Toolbar.jsx
Comment (GuillaumeAmat) [inline src/components/Toolbar/Toolbar.jsx]: The Toolbar did not appear with a transition before. I think it's a bit weird to have a transition on that element right after the page is loaded...

File: stories/Modal.jsx
Comment (GuillaumeAmat) [inline stories/Modal.jsx]: I think we should go for a Modal.Title component. Don't you think? We can't stick with the Sidebar one and I'm sure there will be some differences between the two titles behaviors.

File: src/components/Sidebar/Sidebar.jsx
Comment (GuillaumeAmat) [inline src/components/Sidebar/Sidebar.jsx:73]: We can't close the Sidebar anymore.

**Review by GuillaumeAmat — CHANGES_REQUESTED** [review]

File: stories/Modal.jsx
Comment (bleucitron) [reply stories/Modal.jsx]: I do agree, but I will do it in a dedicated PR.
The `Modal` modifs here were only about using `Transition` instead of `CSSTransitionGroup`.

File: stories/Modal.jsx
Comment (bleucitron) [reply stories/Modal.jsx]: Instead of `padding: 10%;` ?
I have a general feeling that fixing values as `px` is a bad idea, since it's anti-responsive by definition...
This is something I wanted to ask you.
Shouldn't we be using only `rem` or `%` instead of `px` ?

File: src/index.js
Comment (bleucitron) [reply src/index.js]: It is now exported within `src/constants/index.js`.
I thought it was more consistent to have all constants in a dedicated folder.

File: src/components/Toolbar/Toolbar.jsx
Comment (bleucitron) [reply src/components/Toolbar/Toolbar.jsx]: Right, my bad.

File: src/components/Sidebar/Sidebar.jsx
Comment (bleucitron) [reply src/components/Sidebar/Sidebar.jsx:73]: Ok I get it
By the way, why did you apply `appear` on `Sidebar` but not on `Toolbar` ?

**Comment by bleucitron** [comment]
I have fixed the major glitches (this was an unexpected consequence of some small changes that I thought minor).
But the CTA glitch that you pointed out is actually not a glitch. This is just that I haven't styled for small screens yet, since I did not really know what you expected in terms of UX for notifications on mobile. For instance, on mobile notifications should only be with `top` or `bottom` positions in my opinion.

Moreover, I find it not easy to use responsive tools with the storybook environment... Do you have tips on that matter ?

File: src/components/Modal/Modal.jsx
Comment (bleucitron) [reply src/components/Modal/Modal.jsx:123]: There is the same kind of code in `Sidebar`, from which I wrote `Modal`.
In the `<header>`, we display both the title and the header.
We might add the following condition to display the `<header>`: `if (header || title)`.

What do you think ?

File: stories/Modal.jsx
Comment (bleucitron) [reply stories/Modal.jsx]: Actually, I would go for a common set of `Title`, `Header`, `Footer` components, but in their dedicated folder.
The only difference between the behaviors I can think of is about style, but we can totally use `props.classNames` to differentiate. And any kind of `click` behavior can also be set by `props.onClick`.
We should discuss this in a dedicated issue.

File: src/components/Modal/Modal.jsx
Comment (GuillaumeAmat) [reply src/components/Modal/Modal.jsx:123]: In the Sidebar the header contains another components: back button, close button

They can be hidden but it's kind of a convention to reserve the space for the buttons and title even when there is nothing. The main reason is to keep the first readable line at the same position, the eye is not lost between two Sidebars when you navigate through them.

The Modal is different in my opinion. It's a one shot display and I can't see the point to reserve the space, especially on mobile. Try to display the Modal on responsive/mobile mode and you'll see a big empty space on top of the Modal. It's kind of weird (just add `iframe.html` between `localhost/` and the query string to get rid of the Storybook UI ;) ).

File: src/components/Sidebar/Sidebar.jsx
Comment (GuillaumeAmat) [reply src/components/Sidebar/Sidebar.jsx:73]: Sidebars are meant to be shown and hidden. They are not part of the "hard" UI such as the main content and the Toolbars which are fixed (the Toolbar collapse mode is an exception and there is always a fixed button).

For me, Sidebar, Modal and Notification are ephemeral components.

File: src/index.js
Comment (GuillaumeAmat) [reply src/index.js]: Oh you're right, my bad! ^^

File: stories/Modal.jsx
Comment (GuillaumeAmat) [reply stories/Modal.jsx]: Yep you're right for the another-PR point. I did put a comment on the related issue. Thanks!

File: stories/Modal.jsx
Comment (GuillaumeAmat) [reply stories/Modal.jsx]: You're totally right, I missed that. That's said, I really think we should add some space here. We could gain some clarity.

So go for the `rem`! :)

**Comment by GuillaumeAmat** [comment]
> I have fixed the major glitches (this was an unexpected consequence of some small changes that I thought minor).
> But the CTA glitch that you pointed out is actually not a glitch. This is just that I haven't styled for small screens yet, since I did not really know what you expected in terms of UX for notifications on mobile. For instance, on mobile notifications should only be with top or bottom positions in my opinion.

The least point in my opinion is to set `max-width` to 100% on mobile to be sure to display all the content. That's said, I don't really know what UX I prefer:
* Keep it that way and let the Notification appear from the left/top or bottom/right
* Force the Notification to appear from the top/bottom

Maybe the second one? What do you think about that?

> Moreover, I find it not easy to use responsive tools with the storybook environment... Do you have tips on that matter ?

Yep, look at the `iframe.html` tip above ;)

File: src/components/Sidebar/Sidebar.jsx
Comment (bleucitron) [reply src/components/Sidebar/Sidebar.jsx:73]: Okay.
One thing though.
The `_triggerCallback()` function has no real use with the `onExited` prop of `Transition`. Is it okay if I remove it ?

I need to add that I'm not sure what your intention with the `Toolbar` is. I understand that you say that it is of the same kind as `Modal` and `Notification`, which are mounted only when necessary, but the code in `componentWillReceiveProps` seems to say otherwise: 
```
if (this.props.opened !== nextProps.opened) {
      if (nextProps.opened === true) {
        this._triggerCallback('onOpen');
      } else {
        this._triggerCallback('onClose');
      }
    }
```
Actually, if the component is not permanent, its `state.opened` is only used to trigger the closing transition, before it is unmounted. Otherwise, `onOpen` and `onClose` should only be called respectively on `componentDidMount` and `componentWillUnmount`.
In my opinion, there is no need for the code above.

File: src/components/Modal/Modal.jsx
Comment (bleucitron) [reply src/components/Modal/Modal.jsx:123]: Alright, I'll adapt `Modal` in that sense.

File: stories/Modal.jsx
Comment (bleucitron) [reply stories/Modal.jsx]: I'll do that, but will also open an issue for a full `px` to `rem` migration.

**Comment by bleucitron** [comment]
Maybe it can stay that way on mobile, provided the width is 100%. The position will not matter really.
I adjusted a little the style, tell me what you think.

File: src/components/Sidebar/Sidebar.jsx
Comment (GuillaumeAmat) [reply src/components/Sidebar/Sidebar.jsx:73]: I'm not sure to understand your point. So I kind of agree with you if you want to clean the code ;)

I just want to be sure that the call back props will still be present and working though. Is it the case?

Thanks!

File: stories/Modal.jsx
Comment (GuillaumeAmat) [reply stories/Modal.jsx]: Thanks!

**Comment by GuillaumeAmat** [comment]
I'm ok with that! I will test it on real conditions when I will pimp the style after the merge anyway.

File: src/components/Sidebar/Sidebar.jsx
Comment (bleucitron) [reply src/components/Sidebar/Sidebar.jsx:73]: Do not merge the PR now, I will submit a modification for this, and we'll discuss it before merging.

**Comment by GuillaumeAmat** [comment]
I just pushed some commits on `develop` and it generates come conflicts with your PR. Sorry about that ^^

I added the support of the NODE_PATH environment variable so we can/should now use imports without relative path.

**Comment by bleucitron** [comment]
You have changed the location of `colors.js` to the folder `src/themes`. Mine was in `src/constants`. I'm ok with both, but you prefer to have it in `themes`, is that right ?

File: src/components/Alert/Alert.jsx
Comment (GuillaumeAmat) [inline src/components/Alert/Alert.jsx:6]: The contexts are different from one componant to another, I prefer to have it back in the componant file.

But it may be great to export it everytime. That data could be used in the parent project.

**Comment by GuillaumeAmat** [comment]
Hmmm... I don't remember me doing that... But you make a point, it could be in the `constants` folder. Also you made me remember of a comment about contexts that I missed earlier. I just added it.

**Comment by bleucitron** [comment]
Ok, for `colors`, I just noticed you didn't change anything, but this was the previous state of the code before I changed it. So I'll fix the conflict to have `colors` in the `constants` folder.

**Comment by bleucitron** [comment]
Hey @GuillaumeAmat, can you check if the PR is mergeable (which I believe) ?

**Comment by GuillaumeAmat** [comment]
Hey @iOiurson, you asked to not merge the PR here: https://github.com/osm-ui/react/pull/18#discussion_r157352292

Is it ok on your side? I'll merge it if it is ;)

File: src/components/Sidebar/Sidebar.jsx
Comment (bleucitron) [reply src/components/Sidebar/Sidebar.jsx:73]: Okay, I must apologize.

> I need to add that I'm not sure what your intention with the Toolbar is.

There is a typo, I meant "with the Sidebar", which is mounted and dismounted (unlike the `Toolbar`).

The chunk of code I quoted:
```js
if (this.props.opened !== nextProps.opened) {
      if (nextProps.opened === true) {
        this._triggerCallback('onOpen');
      } else {
        this._triggerCallback('onClose');
      }
}
```
was taken from the `Sidebar` component, as all the comments I made on the subject.

I cleaned the `Sidebar` code to get rid of useless `_triggerCallback()`.
Check if the corresponding code fits your expectations.

**Review by GuillaumeAmat — APPROVED** [review]

File: src/components/Sidebar/Sidebar.jsx
Comment (GuillaumeAmat) [reply src/components/Sidebar/Sidebar.jsx:73]: Thank you!

**Comment by GuillaumeAmat** [comment]
Thanks again for your hard work @iOiurson o/
Consider creating your branches directly on that repo in the future. It'll be easier to contribute for me.

File: src/components/Alert/Alert.jsx
Comment (bleucitron) [reply src/components/Alert/Alert.jsx:6]: You merged the PR in spite of this being pending. Should we open an issue about it ?

Comment (GuillaumeAmat) [reply src/components/Alert/Alert.jsx:6]: Oh crap... Yeah, issue or commit as you wish thank you!

**Pull request merged**
